### PR TITLE
Add ability in CNS to setup networking for a docker container

### PR DESCRIPTION
**Pull request by jaer-tsun**
**What this PR does / why we need it**:
This PR
1. Adds CRUD APIs to configure networking for containers via CNS.
2. Supports Azure BATCH as an orchestrator type.

### CODE REVIEW

File: cns/NetworkContainerContract.go [hunk 1]
```diff
@@ -46,6 +51,12 @@ type CreateNetworkContainerRequest struct {
 	Routes                     []Route
 }
 
+// ConfigureNetworkContainerNetworkingRequest - specifies request to attach/detach network container to network.
+type ConfigureNetworkContainerNetworkingRequest struct {
+	DockerContainerid string
```
Comment (tamilmani1989) [inline cns/NetworkContainerContract.go]: I caps

File: cns/NetworkContainerContract.go [hunk 2]
```diff
@@ -46,6 +51,12 @@ type CreateNetworkContainerRequest struct {
...
+	Containerid       string
```
Comment (tamilmani1989) [inline cns/NetworkContainerContract.go]: I caps

File: cns/networkcontainers/networkcontainers.go [hunk 3]
```diff
@@ -57,3 +88,104 @@ func (cn *NetworkContainers) Delete(networkContainerID string) error {
 	log.Printf("[Azure CNS] NetworkContainers.Delete finished.")
 	return err
 }
+
+// This function gets the flattened network configuration (compliant with azure cni) in bytes array format
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: byte array

File: cns/networkcontainers/networkcontainers.go [hunk 4]
```diff
@@ -57,3 +88,104 @@ func (cn *NetworkContainers) Delete(networkContainerID string) error {
...
+			return &emsg
+		}
+	}
+
+	return err
+}
+
+func execPlugin(rt *libcni.RuntimeConf, netconf []byte, operation, path string) error {
+	switch operation {
+	case "ADD":
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: doesn't case statements need to be indented?

File: cns/networkcontainers/networkcontainers.go [hunk 5]
```diff
@@ -4,19 +4,50 @@
...
+	"os/exec"
 
 	"github.com/Azure/azure-container-networking/cns"
 	"github.com/Azure/azure-container-networking/log"
+	"github.com/containernetworking/cni/libcni"
+	"github.com/containernetworking/cni/pkg/invoke"
+)
+
+const (
+	VersionStr              = "cniVersion"
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: why constants are global? Are we using this in some other package?

File: cns/networkcontainers/networkcontainers.go [hunk 6]
```diff
@@ -57,3 +88,104 @@ func (cn *NetworkContainers) Delete(networkContainerID string) error {
...
+
+	var configMap map[string]interface{}
+	if err = json.Unmarshal(content, &configMap); err != nil {
+		log.Printf("[Azure CNS] Failed to unmarshal network configuration with error %v", err)
+		return nil, err
+	}
+
+	// Get the plugins section
+	pluginsSection := configMap[PluginsStr].([]interface{})
+	flatNetConfigMap := pluginsSection[0].(map[string]interface{})
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: can crash..if pluginsSection is empty. check if empty and then call

File: cns/networkcontainers/networkcontainers.go [hunk 7]
```diff
@@ -57,3 +88,104 @@ func (cn *NetworkContainers) Delete(networkContainerID string) error {
...
+		environ := args(operation, path, rt).AsEnv()
+		log.Printf("[Azure CNS] CNI called with environ variables %v", environ)
+		stdout := &bytes.Buffer{}
+		command := exec.Command(path + string(os.PathSeparator) + "azure-vnet")
+		command.Env = environ
+		command.Stdin = bytes.NewBuffer(netconf)
+		command.Stdout = stdout
+		command.Stderr = os.Stderr
+		return pluginErr(command.Run(), stdout.Bytes())
+
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: remove blank

File: cns/networkcontainers/networkcontainers_linux.go [hunk 8]
```diff
@@ -13,10 +21,65 @@ func setWeakHostOnInterface(ipAddress string) error {
...
+func updateInterface(createNetworkContainerRequest cns.CreateNetworkContainerRequest, netpluginConfig *NetPluginConfiguration) error {
+	log.Printf("[Azure CNS] update interface operation called.")
+
+	// Currently update via CNI is only supported for ACI type
+	if createNetworkContainerRequest.NetworkContainerType != cns.AzureContainerInstance {
+		log.Printf("[Azure CNS] operation is only supported for AzureContainerInstance types.")
+		return nil
+	}
+
+	if _, err := os.Stat(netpluginConfig.path); err != nil {
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers_linux.go]: check if netpluginconfig is nil before dereferencing
if netpluginConfig == nil {
    log(errmsg) 
    return err
}

File: cns/networkcontainers/networkcontainers_linux.go [hunk 9]
```diff
@@ -13,10 +21,65 @@ func setWeakHostOnInterface(ipAddress string) error {
...
+	}
+
+	var podInfo cns.KubernetesPodInfo
+	err := json.Unmarshal(createNetworkContainerRequest.OrchestratorContext, &podInfo)
+	if err != nil {
+		log.Printf("[Azure CNS] Unmarshalling %s failed with error %v", createNetworkContainerRequest.NetworkContainerType, err)
+		return err
+	}
+
+	log.Printf("[Azure CNS] Going to update networkign for the pod with Pod info %+v", podInfo)
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers_linux.go]: networking typo

File: cns/networkcontainers/networkcontainers.go [hunk 10]
```diff
@@ -57,3 +88,104 @@ func (cn *NetworkContainers) Delete(networkContainerID string) error {
...
+	return err
+}
+
+func execPlugin(rt *libcni.RuntimeConf, netconf []byte, operation, path string) error {
+	switch operation {
+	case "ADD":
+		fallthrough
+	case "DEL":
+		fallthrough
+	case "UPDATE":
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: add ADD, DEL, UPDATE to constants

File: cns/restserver/restserver.go [hunk 11]
```diff
@@ -1277,3 +1324,129 @@ func (service *HTTPRestService) restoreNetworkState() error {
...
+	}
+
+	if req.DockerContainerid == "" {
+		returnCode = DockerContainerNotSpecified
+		returnMessage = fmt.Sprintf("[Azure CNS] Error. DockerContainerid is empty")
+	} else if req.Containerid == "" {
+		returnCode = NetworkContainerNotSpecified
+		returnMessage = fmt.Sprintf("[Azure CNS] Error. Containerid is empty")
+	}
+
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: why are we continuing to switch if both are empty? Is that valid?

File: cns/restserver/restserver.go [hunk 12]
```diff
@@ -1277,3 +1324,129 @@ func (service *HTTPRestService) restoreNetworkState() error {
...
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: same here..why are we going to switch statement if  both are empty?

File: cns/NetworkContainerContract.go [hunk 13]
```diff
@@ -46,6 +51,12 @@ type CreateNetworkContainerRequest struct {
 	Routes                     []Route
 }
 
+// ConfigureNetworkContainerNetworkingRequest - specifies request to attach/detach network container to network.
```
Comment (tamilmani1989) [inline cns/NetworkContainerContract.go]: whats the difference between dockercontainerid and containerid? Is containerid not the docker container?

Comment (jaer-tsun) [reply cns/NetworkContainerContract.go]: containerid is from swift/dnc

File: cns/NetworkContainerContract.go (hunk 1 above)
Comment (jaer-tsun) [reply cns/NetworkContainerContract.go]: Following the style from CreateNetworkContainerRequest struct

File: cns/NetworkContainerContract.go (hunk 2 above)
Comment (jaer-tsun) [reply cns/NetworkContainerContract.go]: Following the style from CreateNetworkContainerRequest struct

File: cns/networkcontainers/networkcontainers.go (hunk 4 above)
Comment (jaer-tsun) [reply cns/networkcontainers/networkcontainers.go]: not in go style

File: cns/NetworkContainerContract.go (hunk 13 above)
Comment (tamilmani1989) [reply cns/NetworkContainerContract.go]: can we name this as swiftcontainerid then?

File: cns/networkcontainers/networkcontainers.go [hunk 14]
```diff
@@ -57,3 +91,105 @@ func (cn *NetworkContainers) Delete(networkContainerID string) error {
...
+	var configMap map[string]interface{}
+	if err = json.Unmarshal(content, &configMap); err != nil {
+		log.Printf("[Azure CNS] Failed to unmarshal network configuration with error %v", err)
+		return nil, err
+	}
+
+	// Get the plugins section
+	var flatNetConfigMap map[string]interface{}
+	if pluginsSection, ok := configMap[pluginsStr]; ok {
+		flatNetConfigMap = pluginsSection.([]interface{})[0].(map[string]interface{})
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: can't pluginSection be empty array? do we need to check if len(pluginsSection) >0 here..? if you feel its not needed then mark this as resolved. i'm not stressing it

**Comment by tamilmani1989** [comment]
lgtm other than unit tests. if @sharmasushant is fine with it, merge this PR. I'm approving it.

**Pull request merged**